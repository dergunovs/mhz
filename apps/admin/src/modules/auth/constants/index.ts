export const API_POST_LOGIN = 'auth/login';
export const API_GET_CHECK_AUTH = 'auth/check';

export const URL_LOGIN = '/';
