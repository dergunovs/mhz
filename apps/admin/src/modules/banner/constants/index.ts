export const URL_BANNER = '/banners';
export const URL_BANNER_CREATE = `${URL_BANNER}/create`;
export const URL_BANNER_EDIT = `${URL_BANNER}/edit`;
