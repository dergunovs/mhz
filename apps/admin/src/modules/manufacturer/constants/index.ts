export const URL_MANUFACTURER = '/manufacturers';
