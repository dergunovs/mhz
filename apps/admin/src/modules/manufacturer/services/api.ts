import { ComputedRef, Ref } from 'vue';

import { api, IPageQuery, convertParams } from 'mhz-helpers';
import { API_MANUFACTURER, IBaseReply, IManufacturer } from 'mhz-contracts';

export async function getManufacturersApi(query: Ref<IPageQuery | number>) {
  const params = convertParams(query);

  const { data } = await api.get<{ data: IManufacturer[]; total: number }>(API_MANUFACTURER, { params });

  return data;
}

export async function getManufacturerApi(id?: ComputedRef<string | string[]>) {
  if (!id?.value) return null;

  const { data } = await api.get<{ data: IManufacturer }>(`${API_MANUFACTURER}/${id.value}`);

  return data.data;
}

export async function postManufacturerApi(formData: IManufacturer) {
  const { data } = await api.post<IBaseReply>(API_MANUFACTURER, formData);

  return data;
}

export async function updateManufacturerApi(formData: IManufacturer) {
  const { data } = await api.patch<IBaseReply>(`${API_MANUFACTURER}/${formData._id}`, formData);

  return data;
}

export async function deleteManufacturerApi(id?: string) {
  if (!id) return null;

  const { data } = await api.delete<IBaseReply>(`${API_MANUFACTURER}/${id}`);

  return data;
}
