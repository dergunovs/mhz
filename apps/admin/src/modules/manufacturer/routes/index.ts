import { URL_MANUFACTURER, URL_MANUFACTURER_CREATE, URL_MANUFACTURER_EDIT } from '@/manufacturer/constants';

export const manufacturerRoutes = [
  {
    path: URL_MANUFACTURER,
    name: 'ManufacturerList',
    component: () => import('@/manufacturer/pages/ManufacturerListPage.vue'),
  },
  {
    path: URL_MANUFACTURER_CREATE,
    name: 'ManufacturerCreate',
    component: () => import('@/manufacturer/pages/ManufacturerCreatePage.vue'),
  },
  {
    path: `${URL_MANUFACTURER_EDIT}/:manufacturer`,
    name: 'ManufacturerEdit',
    component: () => import('@/manufacturer/pages/ManufacturerEditPage.vue'),
  },
];
