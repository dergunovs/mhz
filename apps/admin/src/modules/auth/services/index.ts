import { Ref } from 'vue';
import { useMutation, useQuery } from '@tanstack/vue-query';

import { ILoginFormData } from '@/auth/interface';
import { API_GET_CHECK_AUTH, API_POST_LOGIN } from '@/auth/constants';

import { api } from '@/common/services/api';

export function postLogin(formData: Ref<ILoginFormData>, options: object) {
  async function fn(): Promise<string | undefined> {
    const { data } = await api.post(API_POST_LOGIN, formData.value);

    return data.token;
  }

  return useMutation({
    mutationKey: [API_POST_LOGIN, formData],
    mutationFn: fn,
    ...options,
  });
}

export function getCheckAuth(options: object) {
  async function fn() {
    const { data } = await api.get(API_GET_CHECK_AUTH);

    return data;
  }

  return useQuery({
    queryKey: [API_GET_CHECK_AUTH],
    queryFn: fn,
    ...options,
  });
}
