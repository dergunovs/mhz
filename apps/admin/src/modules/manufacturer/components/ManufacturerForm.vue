<template>
  <form
    @submit.prevent="props.manufacturer?._id ? update() : submit()"
    :class="$style.form"
    data-test="manufacturer-form"
  >
    <UiField label="Title" isRequired :error="error('title')">
      <UiInput v-model="formData.title" isFocus data-test="manufacturer-form-title" />
    </UiField>

    <UiField label="Description" isRequired :error="error('description')">
      <UiEditor v-model="formData.description" data-test="manufacturer-form-description" />
    </UiField>

    <UiField label="Country" isRequired :error="error('country')">
      <UiSelect
        v-model="formData.country"
        :options="countries"
        isFilter
        lang="en"
        data-test="manufacturer-form-country"
      />
    </UiField>

    <UiUpload
      label="Upload logo"
      :file="logoFile"
      isSingle
      isRequired
      :isDisabled="!!formData.logoUrl"
      :error="error('logoUrl')"
      :extensions="['png']"
      @add="addLogoFile"
      @remove="removeLogoFile"
      @upload="mutateUploadFile(logoFile)"
      lang="en"
    />

    <ImagePreview
      v-if="formData.logoUrl"
      :urls="[formData.logoUrl]"
      @delete="formData.logoUrl = ''"
      data-test="manufacturer-form-logo"
    />

    <FormButtons
      :id="props.manufacturer?._id"
      :isLoading="isLoadingPost || isLoadingUpdate"
      @delete="handleDelete"
      data-test="manufacturer-form-buttons"
    />
  </form>
</template>

<script setup lang="ts">
import { ref, onMounted } from 'vue';
import { useRouter } from 'vue-router';

import { UiField, UiInput, UiUpload, toast, UiSelect, UiEditor } from 'mhz-ui';
import { countries } from 'mhz-countries';
import { clone, useValidator, required, useQueryClient } from 'mhz-helpers';
import { API_MANUFACTURER, IManufacturer } from 'mhz-contracts';

import ImagePreview from '@/common/components/ImagePreview.vue';
import FormButtons from '@/common/components/FormButtons.vue';

import { URL_MANUFACTURER } from '@/manufacturer/constants';
import { postManufacturer, updateManufacturer, deleteManufacturer } from '@/manufacturer/services';
import { uploadFile } from '@/common/services';

interface IProps {
  manufacturer?: IManufacturer;
}

const props = defineProps<IProps>();

const queryClient = useQueryClient();

const router = useRouter();

const formData = ref<IManufacturer>({
  title: '',
  description: '',
  logoUrl: '',
  country: '',
});

const { mutate: mutatePost, isPending: isLoadingPost } = postManufacturer({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_MANUFACTURER] });
    toast.success('Manufacturer added');
    router.push(URL_MANUFACTURER);
  },
});

const { mutate: mutateUpdate, isPending: isLoadingUpdate } = updateManufacturer({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_MANUFACTURER] });
    toast.success('Manufacturer updated');
  },
});

const { mutate: mutateDelete } = deleteManufacturer({
  onSuccess: async () => {
    queryClient.removeQueries({ queryKey: [API_MANUFACTURER] });
    await queryClient.refetchQueries({ queryKey: [API_MANUFACTURER] });
    toast.success('Manufacturer deleted');
    router.push(URL_MANUFACTURER);
  },
});

const { error, isValid } = useValidator(formData, {
  title: [required('en')],
  description: [required('en')],
  logoUrl: [required('en')],
  country: [required('en')],
});

function submit() {
  if (isValid()) mutatePost(formData.value);
}

function update() {
  if (isValid()) mutateUpdate(formData.value);
}

function handleDelete(id: string) {
  mutateDelete(id);
}

const logoFile = ref<File>();

function addLogoFile(file: File) {
  logoFile.value = file;
}

function removeLogoFile() {
  logoFile.value = undefined;
}

const { mutate: mutateUploadFile } = uploadFile({
  onSuccess: (data: string) => {
    formData.value.logoUrl = data;
    removeLogoFile();
    toast.success('Logo added');
  },
});

onMounted(() => {
  if (props.manufacturer) formData.value = clone(props.manufacturer);
});
</script>

<style module lang="scss">
.form {
  display: flex;
  flex-direction: column;
  gap: 24px;
  align-items: flex-start;
}
</style>
