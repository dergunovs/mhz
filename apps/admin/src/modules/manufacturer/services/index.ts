import { useMutation, useQuery } from '@tanstack/vue-query';

import { IManufacturer } from 'mhz-types';

import { API_MANUFACTURER } from '@/manufacturer/constants';
import { api } from '@/common/services/api';

export function getManufacturers() {
  async function fn(): Promise<IManufacturer[] | undefined> {
    const { data } = await api.get(API_MANUFACTURER);

    return data;
  }

  return useQuery({
    queryKey: [API_MANUFACTURER],
    queryFn: fn,
  });
}

export function postManufacturer(options: object) {
  async function fn(formData: IManufacturer) {
    await api.post(API_MANUFACTURER, formData);
  }

  return useMutation({
    mutationKey: [API_MANUFACTURER],
    mutationFn: fn,
    ...options,
  });
}

export function deleteManufacturer(options: object) {
  async function fn(id?: number) {
    await api.delete(`${API_MANUFACTURER}/${id}`);
  }

  return useMutation({
    mutationKey: [API_MANUFACTURER],
    mutationFn: fn,
    ...options,
  });
}
