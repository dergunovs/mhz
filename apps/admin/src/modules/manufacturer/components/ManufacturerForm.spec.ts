import { DefineComponent } from 'vue';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount, flushPromises } from '@vue/test-utils';
import { IManufacturer, IBaseReply, API_MANUFACTURER } from 'mhz-contracts';
import { toast } from 'mhz-ui';
import * as helpers from 'mhz-helpers';
import { dataTest } from 'mhz-helpers';

import ManufacturerForm from './ManufacturerForm.vue';

import { wrapperFactory, mockMutationReply, router } from '@/common/test';
import { MANUFACTURERS } from '@/manufacturer/fixtures';
import { URL_MANUFACTURER } from '@/manufacturer/constants';
import * as manufacturerServices from '@/manufacturer/services';

const MANUFACTURER = MANUFACTURERS.data[0];

let onSuccessPost: () => void;
const spyMutatePost = vi.fn();

vi.spyOn(manufacturerServices, 'postManufacturer').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessPost = options.onSuccess;

  return mockMutationReply<IBaseReply, IManufacturer>(spyMutatePost);
});

let onSuccessUpdate: () => void;
const spyMutateUpdate = vi.fn();

vi.spyOn(manufacturerServices, 'updateManufacturer').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessUpdate = options.onSuccess;

  return mockMutationReply<IBaseReply, IManufacturer>(spyMutateUpdate);
});

let onSuccessDelete: () => void;
const spyMutateDelete = vi.fn();

vi.spyOn(manufacturerServices, 'deleteManufacturer').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessDelete = options.onSuccess;

  return mockMutationReply<IBaseReply, string | undefined>(spyMutateDelete);
});

const spyRefetchQueries = vi.fn();
const spyRemoveQueries = vi.fn();

vi.spyOn(helpers, 'useQueryClient').mockReturnValue({
  refetchQueries: spyRefetchQueries,
  removeQueries: spyRemoveQueries,
} as unknown as helpers.QueryClient);

const spyRouterPush = vi.spyOn(router, 'push');
const spyToastSuccess = vi.spyOn(toast, 'success');

const QUERY_KEY = { queryKey: [API_MANUFACTURER] };

const manufacturerForm = dataTest('manufacturer-form');
const manufacturerFormTitle = dataTest('manufacturer-form-title');
const manufacturerFormDescription = dataTest('manufacturer-form-description');
const manufacturerFormCountry = dataTest('manufacturer-form-country');
const manufacturerFormLogo = dataTest('manufacturer-form-logo');
const manufacturerFormButtons = dataTest('manufacturer-form-buttons');

let wrapper: VueWrapper;

beforeEach(() => {
  wrapper = wrapperFactory(ManufacturerForm, {
    props: {
      manufacturer: MANUFACTURER,
    },
  });
});

enableAutoUnmount(afterEach);

describe('ManufacturerForm', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(manufacturerForm)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shows manufacturer content in fields if props is passed', async () => {
    expect(wrapper.findComponent(manufacturerFormTitle).attributes('modelvalue')).toBe(MANUFACTURER.title);
    expect(wrapper.findComponent(manufacturerFormDescription).attributes('modelvalue')).toBe(MANUFACTURER.description);
    expect(wrapper.findComponent(manufacturerFormCountry).attributes('modelvalue')).toBe(MANUFACTURER.country);
    expect(wrapper.findComponent(manufacturerFormLogo).attributes('urls')).toBe(MANUFACTURER.logoUrl);
  });

  it('shows icon preview if manufacturer has fields', async () => {
    expect(wrapper.findComponent(manufacturerFormLogo).exists()).toBe(true);

    await wrapper.setProps({ manufacturer: undefined });

    expect(wrapper.findComponent(manufacturerFormLogo).exists()).toBe(true);
  });

  it('updates manufacturer', async () => {
    expect(spyMutateUpdate).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);

    const NEW_MANUFACTURER_TITLE = 'yo';

    await wrapper.findComponent(manufacturerFormTitle).setValue(NEW_MANUFACTURER_TITLE);

    await wrapper.find(manufacturerForm).trigger('submit');

    expect(spyMutateUpdate).toBeCalledTimes(1);
    expect(spyMutateUpdate).toBeCalledWith({ ...MANUFACTURER, title: NEW_MANUFACTURER_TITLE });

    onSuccessUpdate();

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);
  });

  it('deletes manufacturer', async () => {
    expect(spyMutateDelete).toBeCalledTimes(0);
    expect(spyRemoveQueries).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyRouterPush).toBeCalledTimes(0);

    wrapper.findComponent<DefineComponent>(manufacturerFormButtons).vm.$emit('delete', MANUFACTURER._id);

    expect(spyMutateDelete).toBeCalledTimes(1);
    expect(spyMutateDelete).toBeCalledWith(MANUFACTURER._id);

    onSuccessDelete();

    expect(spyRemoveQueries).toBeCalledTimes(1);
    expect(spyRemoveQueries).toBeCalledWith(QUERY_KEY);

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_MANUFACTURER);
  });

  it('creates manufacturer', async () => {
    wrapper.unmount();

    expect(spyMutatePost).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyRouterPush).toBeCalledTimes(0);

    const wrapperWithoutManufacturer = wrapperFactory(ManufacturerForm, {});

    const NEW_MANUFACTURER = {
      title: 'hi',
      description: 'text',
      country: 'Russia',
      logoUrl: 'pic.jpg',
    };

    await wrapperWithoutManufacturer.findComponent(manufacturerFormTitle).setValue(NEW_MANUFACTURER.title);
    await wrapperWithoutManufacturer.findComponent(manufacturerFormDescription).setValue(NEW_MANUFACTURER.description);
    await wrapperWithoutManufacturer.findComponent(manufacturerFormCountry).setValue(NEW_MANUFACTURER.country);

    wrapperWithoutManufacturer.vm.formData.logoUrl = NEW_MANUFACTURER.logoUrl;

    await wrapperWithoutManufacturer.find(manufacturerForm).trigger('submit');

    expect(spyMutatePost).toBeCalledTimes(1);
    expect(spyMutatePost).toBeCalledWith(NEW_MANUFACTURER);

    onSuccessPost();

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_MANUFACTURER);
  });
});
