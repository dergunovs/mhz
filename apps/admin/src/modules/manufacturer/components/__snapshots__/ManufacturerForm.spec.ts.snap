// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ManufacturerForm > matches snapshot 1`] = `
"<form class="_form_df4d98" data-test="manufacturer-form">
  <fieldset label="Title" isrequired=""><input type="text" modelvalue="Apacer" isfocus="" data-test="manufacturer-form-title"></fieldset>
  <fieldset label="Description" isrequired=""><input type="text" data-test="manufacturer-form-description"></fieldset>
  <fieldset label="Country" isrequired=""><input type="text" modelvalue="China" options="Abkhazia,Australia,Austria,Azerbaijan,Aland Islands,Albania,Algeria,Anguilla,Angola,Andorra,Argentina,Armenia,Aruba,Afghanistan,Bahamas,Bangladesh,Barbados,Bahrain,Belarus,Belize,Belgium,Benin,Bulgaria,Bolivia,Bosnia &amp; Herzegovina,Botswana,Brazil,Brunei Darussalam,Burundi,Bhutan,Vatican City,United Kingdom,Hungary,Venezuela,Timor, East,Viet Nam,Gabon,Haiti,Gambia,Ghana,Guadeloupe,Guatemala,Guinea,Guinea-Bissau,Germany,Gibraltar,Hong Kong,Honduras,Grenada,Greenland,Greece,Georgia,Guam,Denmark,Dominica,Dominican Republic,Egypt,Zambia,Western Sahara,Zimbabwe,Israel,India,Indonesia,Jordan,Iraq,Iran,Ireland,Iceland,Spain,Italy,Yemen,Kazakhstan,Cambodia,Cameroon,Canada,Qatar,Kenya,Cyprus,Kyrgyzstan,Kiribati,China,Colombia,Korea, D.P.R.,Korea,Costa Rica,Cote d'Ivoire,Cuba,Kuwait,Lao P.D.R.,Latvia,Lesotho,Liberia,Lebanon,Libyan Arab Jamahiriya,Lithuania,Liechtenstein,Luxembourg,Mauritius,Mauritania,Madagascar,Macedonia,Malawi,Malaysia,Mali,Maldives,Malta,Morocco,Mexico,Mozambique,Moldova,Monaco,Mongolia,Namibia,Nepal,Niger,Nigeria,Netherlands,Nicaragua,New Zealand,Norway,United Arab Emirates,Oman,Pakistan,Panama,Paraguay,Peru,Poland,Portugal,Russia,Romania,San Marino,Saudi Arabia,Senegal,Serbia,Singapore,Syrian Arab Republic,Slovakia,Slovenia,Somalia,Sudan,USA,Tajikistan,Thailand,Tanzania,Togo,Tunisia,Turkmenistan,Turkey,Uganda,Uzbekistan,Ukraine,Uruguay,Micronesia,Fiji,Philippines,Finland,France,Croatia,Chad,Montenegro,Czech Republic,Chile,Switzerland,Sweden,Sri Lanka,Ecuador,Eritrea,Estonia,Ethiopia,South Africa,Jamaica,Japan" isfilter="" lang="en" data-test="manufacturer-form-country"></fieldset><input type="file" label="Upload logo" issingle="" isrequired="" isdisabled="true" extensions="png" lang="en">
  <image-preview-stub urls="resized-1692211814147-Apacer.png" isthumb="false" data-test="manufacturer-form-logo"></image-preview-stub>
  <form-buttons-stub id="64dd1a67f8fd2bdaa35cc735" data-test="manufacturer-form-buttons"></form-buttons-stub>
</form>"
`;
